src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php

```diff
@@ -94,12 +95,32 @@ private function executeGuardAuthenticator($uniqueGuardKey, GuardAuthenticatorIn
                 $this->logger->debug('Calling getCredentials() on guard configurator.', array('firewall_key' => $this->providerKey, 'authenticator' => get_class($guardAuthenticator)));
             }
 
+            // abort the execution of the authenticator if it doesn't supports the request.
+            if ($guardAuthenticator instanceof GuardAuthenticatorInterface) {
+                if (true !== $guardAuthenticator->supports($request)) {
+                    return;
+                }
+                // as there was a support for given request,
+                // authenticator is expected to give not-null credentials.
+                $credentialsCanBeNull = false;
+            } else {
+                // it's a GuardAuthenticatorLegacyInterface
+                // we support the previous behaviour to avoid BC break.
+                $credentialsCanBeNull = true;
+            }
+
             // allow the authenticator to fetch authentication info from the request
             $credentials = $guardAuthenticator->getCredentials($request);
 
-            // allow null to be returned to skip authentication
             if (null === $credentials) {
-                return;
+                // if GuardAuthenticatorLegacyInterface is used
+                // allow null to skip authentication.
+                if ($credentialsCanBeNull) {
+                    return;
+                }
+
+                // otherwise something went wrong and the authentication must fail
+                throw new \InvalidArgumentException('Guard authenticator must provide user credentials');
```

Comment:
A RuntimeException (or even better its child UnexpectedValueException) would be better. This is not an invalid argument here, but an invalid return value